### ISSUE TITLE
Resource.render should allow returning a Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11260]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#253|
|Type|enhancement|
|Created|2003-09-16 03:59:40Z|

Attachments:

* [render-deferred.patch.rtf](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4c1/4c15dc21c91634c1b301de6236eb08ead86be4ae/af4a41a3a726537fef2ead32065f190a106aeb29.rtf) (5508 bytes) - added by jknight on 2003-09-16 03:59:40Z - 
* [render-deferred.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4c1/4c15dc21c91634c1b301de6236eb08ead86be4ae/66d458cc15a5d15fb3248f63dadb5905f97a23e7.patch) (5098 bytes) - added by jknight on 2003-09-16 04:04:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__253 253
type__enhancement enhancement
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__web web
keywords__ 
time__1063684780000000 1063684780000000
changetime__1468062227168355 1468062227168355
version__ 
owner__ 
cc__itamarst cc__dp cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This patch fixes the API for Resource.render to allow you to return a 
Deferred, as one would expect to be able to do. This is as opposed to 
the currently recommended method of returning NOT_DONE_YET and 
restructuring your code to call request.write(...). Of course, that should 
still work, too.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
bleargh, stupid RTF file.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
There are a few places in the code that assume result of
render() is either NOT_DONE_YET or string. This is typically
done in wrapper resources that want to be transparent and
not alter request (e.g. my patch for improving
DeferredResource, or twisted.web.guard.) I suspect this
patch will break them.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
I've removed the [PATCH] indicator, since the current patch 
isn't complete -- like itamar said, at least patch 
twisted.web.guard

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
It's not just web.guard, it's other code in my own projects
and possibly others. I'm not sure if there is a solution :(
You can do stuff like having a reallyRender that render
calls and can return Deferreds which are converted to
NOT_DONE_YET, or... I dunno. Make __eq__ and friends for
NOT_DONE_YET do some magic for deferreds?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Surely t.w.guard should be fixed (sorry didn't notice that problem). 
However, I'm not sure how much effort really needs to be put into being 
totally compatible with user code. Here is my proposed solution:
1) For the next X amount of time, do not return a deferred from render() 
in any code in the twisted framework (unless it's just passing it through 
a user's render()).
2) Make all code in twisted framework able to deal with user code 
returning a deferred.

With this solution, breakage can only occur if a user's code returns a 
deferred from render, *and* the user's code isn't able to deal with that. 
Nothing that currently works would break.

I think that's a reasonable compromise.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
That seems like a reasonable solution, yes. Some sort of
utility code for wrapper resources so they don't have to
write the same code each time would also make it easier to
transition - "oh, just use this if you're doing that sort of
thing".

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Can this be downgraded from "bug" to "feature"?
itamar?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Somehow I feel as though 'bug' is the right classification, but at the same time
it's changing an existing interface which is ANCIENT in Twisted's time-scale
(which has problems as indicated).

jknight, can you update the patch to provide the policy that you suggest?  It
sounds good.

I'm assigning this to dp to make sure someone is in charge of following up and
getting this committed eventually.  I'd like it to be in 1.1 so that we can end
end "X amount of time" when we hit 1.2 and drop the other vestiges of ugly
backwards compatibility.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I've tried to decipher the widget code in t.w.guard. The simple solution of just returning the 
subresource's render(..) isn't possible because I guess the return value of reallyRender needs to be 
widgets.FORGET_IT so that the first time after it authenticates you, it doesn't redisplay the login 
form.

So, anyhow -- I *think* guard can be fixed by the following. It is somewhat odd as it calls back 
into the request's render method which you don't generally do, but it should work because the first 
render call aborted with NOT_DONE_YET and no headers have been sent yet. The test doesn't run 
because DummyRequest doesn't implement render. I don't see a simple example of t.w.guard use
and have no idea how all this set of authorizer/etc stuff works. 

So, this is untested but probably works. :)

@@ -210,10 +207,7 @@
     def reallyRender(self, request):
         # it's authenticated already...
         res = self.res.getChildForRequest(request)
-        val = res.render(request)
-        if val != NOT_DONE_YET:
-            request.write(val)
-            request.finish()
+        request.render(res)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
This won't work for the case returns a string. Or for the case it returns a
Deferred ;)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Hm? Explain? request.render doesn't return anything.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Good point! I thought you did resource.render. I'll check the request.render. If
it *doesn't* do what we want, we ought to add a method to Request that does :)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Re: testing.
The latest tests in test_web.py show a way to use real requests, not dummy
ones.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Now that I've implemented that render_* change, allowing resources to return
deferreds from render_* should be easy: put the code to manage it in base
Resource.render.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
does nevow/twisted.web2 do this already?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This bug is irrelevant now, people should use Nevow, or web2 whenever it's done to get this 
functionality. Closing wontfix.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_18"></a>|@ghazel set status to `reopened`|
|-|-|

twisted.web is alive and well. People very recently have been wondering why this wasn't ever improved:
http://jcalderone.livejournal.com/51109.html?thread=116389#t116389
http://jcalderone.livejournal.com/50226.html?thread=113970#t113970

Perhaps we should revisit this issue?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `closed`|
|-|-|

[#3711](https://github.com/twisted/twisted/issues/3711) was a duplicate of this.  Since a bunch of work has been started there, we'll keep this ticket closed as the duplicate though (whereas normally I would close the higher numbered ticket as the duplicate).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|GitHub commented|
|-|-|

In changeset 177052b0ff1eb140b03c728a1f1caacb431fa998
```
#!CommitTicketReference repository="" revision="177052b0ff1eb140b03c728a1f1caacb431fa998"
Merge pull request #253 from twisted/8527-rodrigc-file

Author: rodrigc
Reviewer: Lukasa, alex
Fixes: #8527

Replace file() with open() in twisted/internet
```
